zephyr/lib/regions_mm.c

```diff
@@ -1,52 +1,717 @@
+			 * 0 for individually allocated block and 0 for block ending allocation
+			 * Continuing the example above
+			 * First: 1111100000000000 allocate another block 1111100000000000
+			 * Second:1110000000000000 after the first one	  1110000000000000
+			 * based on information above we know exactly how long allocations were
+			 * by checking second bit array against first one
+			 * First: 1111111100000000 allocate another block 1111111100000000
+			 * Second:1110000000000000 and another two blocks 1110001000000000
+			 * We are still able to pinpoint allocations size
+			 */
+
+			/* Set bits count - 1 on the same offset as in allocation bit array */
+			allocation_bitarray_offset = (uintptr_t)ptr
+					- (uintptr_t)heap->physical_blocks_allocators
+						[mem_block_iterator]->buffer;
+			allocation_bitarray_position = allocation_bitarray_offset / block_size;
+
+			/* Need to set bits only if we have more than 1 block to allocate */
+			if (block_count - 1)
+				sys_bitarray_set_region(heap->allocation_sizes[mem_block_iterator],
+					block_count - 1, allocation_bitarray_position);
 			break;
-		case MEM_REG_ATTR_OPPORTUNISTIC_MEMORY:
-			vm_heaps[i].memory_caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP;
+		}
+	}
+
+	/* If ptr is NULL it means we did not allocate anything and we should
+	 * break execution here
+	 */
+	if (!ptr)
+		return NULL;
+
+	/* Now we need to check if have mapped physical page already
+	 * We can do it by working out if there was any other allocation
+	 * done in the allocator section that is in CONFIG_MM_DRV_PAGE_SIZE chunk.
+	 * Start by getting start of allocation
+	 * ptr aligned to physical pages boundary
+	 */
+	phys_aligned_ptr = (void *)ALIGN_DOWN((uintptr_t)ptr, CONFIG_MM_DRV_PAGE_SIZE);
+	phys_aligned_alloc_end = (void *)ALIGN_UP((uintptr_t)ptr
+		+ alloc_size, CONFIG_MM_DRV_PAGE_SIZE);
+
+	/* To check if there was any allocation done before on block we need
+	 * to perform check operations however
+	 * we just made allocation there. We will clear it
+	 * from bit array of allocator for ease of operation.
+	 * RFC if someone knows a better way to handle it in bit array API.
+	 */
+
+	/* Clear bits of the allocation that was done */
+	sys_bitarray_clear_region(heap->physical_blocks_allocators[mem_block_iterator]->bitmap,
+		block_count, allocation_bitarray_position);
+
+	/* Now that we cleared the allocation we just made we can check
+	 * if there was any other allocation in the space that would need to be
+	 * mapped. Since this API is only one that uses mapping in specified address
+	 * ranges we can assume that we made mapping for the space if it has any previous
+	 * allocations. We check that iterating over all physical allocation blocks
+	 */
+	physical_block_count = ((uintptr_t)phys_aligned_alloc_end
+		- (uintptr_t)phys_aligned_ptr) / CONFIG_MM_DRV_PAGE_SIZE;
+
+	/* needed in case of mid mapping failure */
+	void *ptrs_to_map[physical_block_count];
```

Comment:
This struct saves current mappings in case of mid mapping failure.  How else would You approach such a case since every mapping is different ? You would need to limit the size of the single allocation. So it is in fact limited by the size of the heap itself - code cannot get to there if allocation of virtual space was not successfully
Also it is supposed to save only ptrs not a complex structures. What makes it so scary ?

Reply:
@dabekjakub The stack space is very scarce resource. We still use quite big stack sizes, but it's going to be hard to optimize this if we have a lot of usages like this code that does dynamic allocations on stack. Given this a single item is only a 32bit address like you note, wouldn't it be safer to allocate the array with worst-case size, and check the physical_block_count does not go beyond this?

Reply:
see here: 
https://github.com/zephyrproject-rtos/zephyr/blob/402d4eb0dab2b5dd81cdc9c2520625f39f5b787f/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c#L236

looks like it is, but limited to SOC_INTEL_COMM_WIDGET "Intel Communication Widget driver"

if you need it, it should be enabled permanently and exposed in TLB driver API

Reply:
@dabekjakub somehow this needs to be bounded. Already a 16 item rollback buffer (for a 64kB alloc) is quite big to store on stack. Given this is only needed on a rare error path, it wouldn't be a problem if the rollback would be slow...

Reply:
@dabekjakub  The solution in updated PR is now ok. Additional heap alloc is not ideal, but not a dealbreaker either.